installer/test/mix/tasks/alice_new_handler_test.exs

```diff
@@ -0,0 +1,131 @@
+Code.require_file("mix_helper.exs", __DIR__)
+
+defmodule Mix.Tasks.Alice.New.HandlerTest do
+  use ExUnit.Case, async: false
+
+  import AliceNew.MixHelper
+  import ExUnit.CaptureIO
+  alias Mix.Tasks.Alice.New.Handler
+
+  @handler_name "super_awesome"
+  @app_name "alice_#{@handler_name}"
+  @module_name "SuperAwesome"
+  @alice_version AliceNew.Utilities.alice_version()
+
+  test "alice.new.handler with defaults" do
+    in_tmp("new handler with defaults", fn ->
+      output =
+        capture_io(fn ->
+          Handler.run([@handler_name])
+
+          assert_file("#{@app_name}/README.md", fn file ->
+            assert file =~ "AliceSuperAwesome"
+            assert file =~ ~s({:alice, "~> #{@alice_version}"})
+          end)
+
+          assert_file("#{@app_name}/.formatter.exs", fn file ->
+            assert file =~ "import_deps: [:alice]"
+          end)
+
+          assert_file("#{@app_name}/.gitignore")
+
+          assert_file("#{@app_name}/config/config.exs")
+
+          assert_file("#{@app_name}/lib/alice/handlers/#{@handler_name}.ex", fn file ->
+            assert file =~ "defmodule Alice.Handlers.#{@module_name} do"
+            assert file =~ "use Alice.Router"
+          end)
+
+          assert_file("#{@app_name}/mix.exs", fn file ->
+            assert file =~ "defmodule Alice#{@module_name}.MixProject do"
+            assert file =~ "app: :#{@app_name}"
+            assert file =~ ~s({:ex_doc, ">= 0.0.0", only: :dev, runtime: false})
+            assert file =~ ~s({:alice, "~> #{@alice_version}"})
+          end)
+
+          assert_file("#{@app_name}/test/test_helper.exs")
+
+          assert_file("#{@app_name}/test/alice/handlers/#{@handler_name}_test.exs", fn file ->
+            assert file =~ "defmodule Alice.Handlers.#{@module_name}Test do"
+            assert file =~ "use Alice.HandlerCase, handlers: Alice.Handlers.SuperAwesome"
+            assert file =~ "send_message"
+            assert file =~ "first_reply"
+          end)
+        end)
+
+      assert output =~ "Your Alice handler was created successfully."
+    end)
+  end
```

Comment:
Should this be a describe block for "new handler with defaults" where each file is a separate test? If for no other reason then to help isolate things when something does break? Something makes me instinctive shy away from a test with this many asserts.

```elixir
describe "when running a new test with defaults" do
  test "it should generate README file" do
    in_tmp("new handler with defaults", fn ->
      output =
        capture_io(fn ->
          Handler.run([@handler_name])

          assert_file("#{@app_name}/README.md", fn file ->
            assert file =~ "AliceSuperAwesome"
            assert file =~ ~s({:alice, "~> #{@alice_version}"})
          end)
        end)
      assert output =~ "Your Alice handler was created successfully."
    end)
  end
end
```
Something like that? Or is that overkill?

Reply:
In ruby it's not good to have more than one assertion per test because when it rails it's difficult to tell which failed. Having multiple assertions per test is pretty common in Elixir because ExUnit does a better job telling you where the failure was.

Reply:
Also in ExUnit, you can't nest describe blocks, so they don't tend to be used as much overall. Rather than being used for organization like is often the case in RSpec etc, in ExUnit they are mostly used to group different tests that need a common setup block.

Reply:
Limiting the number of asserts per test isn’t really a rubyism as a result of tooling, it’s long been a kind of best practice to avoid run on tests or the like. Not to mention the normal better readability that comes from smaller functions with explicit names.

I’m definitely not a purist who would advise for one assert per test, but I do try to only test only one logical concept per test. With push and pull based on complication of setup. This struck me as one of those kinds of tests where I can see clean logical lines in groups of asserts, counterbalanced by setup costs.

Feel free to resolve the conversation if you don’t think it’s applicable. /shrug

Reply:
I realize that about describe blocks. I was thinking those tests all do have a common setup, it’s just being done inside the test now 😊